modules/custom/wim_cvdr/wim_cvdr.module

```diff
@@ -0,0 +1,235 @@
+
+  $items['admin/config/services/cvdr'] = array(
+    'title' => 'WIM CVDR',
+    'description' => 'Configuration for connecting with CVDR.',
+    'page callback' => 'drupal_get_form',
+    'page arguments' => array('_wim_cvdr_admin_page'),
+    'access arguments' => array('administer blocks'),
+    'type' => MENU_NORMAL_ITEM,
+    'file' => 'wim_cvdr.admin.inc',
+  );
+
+  return $items;
+}
+
+/**
+ * Implements hook_block_info().
+ */
+function wim_cvdr_block_info() {
+  $blocks = array();
+  $subjects = _wim_cvdr_retrieve_subjects();
+
+  // Create a seperate block for each subject.
+  foreach ($subjects as $subject) {
+    // Create nice machine name based on subject name.
+    $machine_name = str_replace('-', '_', drupal_html_class($subject));
+    $blocks[md5('wim_cvdr_' . $machine_name)] = array(
+      'info' => t('CVDR: !subject', array('!subject' => $subject)),
+      'cache' => DRUPAL_CACHE_GLOBAL,
+    );
+  }
+
+  return $blocks;
+}
+
+/**
+ * Implements hook_block_view().
+ */
+function wim_cvdr_block_view($delta = '') {
+  $block = array();
+
+  $subjects = _wim_cvdr_retrieve_subjects();
+  foreach ($subjects as $subject) {
+    // Create nice machine name based on subject name.
+    $machine_name = str_replace('-', '_', drupal_html_class($subject));
+
+    // Check if block is current subject.
+    if ($delta === md5('wim_cvdr_' . $machine_name)) {
+      // Render subject block.
+      $block['subject'] = $subject;
+
+      // Build the links to the CVDR items.
+      $items = _wim_cvdr_retrieve_items($subject);
+
+      // Sort the items alphabetically.
+      sort($items);
+
+      foreach ($items as $item) {
+        $list_items[] = l($item['title'], $item['url'],
+          array('attributes' => array('target' => '_blank')));
+      }
+
+      $block['content'] = theme('item_list', array(
+        'items' => $list_items,
+        'title' => '',
+        'type' => 'ul',
+        'attributes' => array(
+          'class' => array('cvdr-wim', drupal_html_class($subject)),
+        ),
+      ));
+    }
+  }
+
+  // Add JS.
+  drupal_add_js(array(
+    'wim_cvdr' => array(
+      'maxItems' => variable_get('wim_cvdr_max_visible', 5) - 1,
+    ),
+  ), 'setting');
+  drupal_add_js(
+    drupal_get_path('module', 'wim_cvdr') . '/wim_cvdr.js', array(
+      'type' => 'file',
+      'scope' => 'footer',
+    )
+  );
+
+  return $block;
+}
+
+/**
+ * Retrieve the subjects available from the webservice.
+ *
+ * @return array
+ *   containing the subjects.
+ */
+function _wim_cvdr_retrieve_subjects() {
+  $items = array();
+
+  // Use caching to improve performance.
+  if ($cached = cache_get('wim_cvdr_items')) {
+    $items = $cached->data;
+  }
+  else {
+    $xml = _wim_cvdr_read_xml();
+    if ($xml) {
+      $items = _wim_cvdr_process_xml($xml);
+    }
+  }
+
+  return array_keys($items);
+}
+
+/**
+ * Retrieve all items from the webservice.
+ *
+ * @param string $key
+ *   Indicating which block to render.
+ *
+ * @return array
```

Comment:
The types returned by this function differ depending on the context. The return type should be the same in all scenarios so that it's easy for the calling code to handle.

E.g. an empty array signals that no results could be found. An array with a single entry provides the request array. An array containing multiple entries can signal all (matching) results.

Always returning an array makes it possible for the calling code to iterate over the result without checking for each scenario individually.